llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2756,6 +2756,147 @@ LogicalResult WinogradOutputTransformOp::reifyResultShapes(
       .reifyResultShapes(b, reifiedReturnShapes);
 }
 
+//===----------------------------------------------------------------------===//
+// FlashAttentionFwdOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult FlashAttentionFwdOp::verify() {
+  Operation *op = getOperation();
+  if (getNumInputs() != 3) {
+    return op->emitOpError("expected three input operands");
+  }
+  if (getNumOutputs() != 1) {
+    return op->emitOpError("expected one output operand");
+  }
```

Comment:
Why these aren't specified in the ODS instead of using variadic?

Reply:
This is just because they implement the `LinalgExtInterface`.... We could probably do it better and reduce what we use the interface for.... But the interface expects `inputs` and `outputs`. It was done to just make a few things easier (short term).

Reply:
The interface just requires that `getInputs` and `getOutputs` methods to exist, it doesn't specify how they are implemented. The ODS part could just have:

```
let extraClassDeclaration = [{
  ValueRange getInputs() { return getOperands().take_front(3); }
  ValueRange getOutputs() { return getOperands().drop_front(3).take_front(); }
}];
```

and have proper accessors and verifiers generated.

Reply:
Makes sense. Thanks! Will incorporate into subsequent patches.

Reply:
Just a follow up on this, the generated file for LinalgExtInterfaces specifies 
```
ValueRange getInputs(const Concept *impl, ::mlir::Operation *tablegen_opaque_val);
```
but 
```
return getOperands().take_front(3);
```
returns a ```mlir::OperandRange```. Do you know how to convert one to the other?

Reply:
ValueRange is implicitly constructible form OperandRange - https://github.com/llvm/llvm-project/blob/5a7f47cc021bd7a19cb70c9a30755d6b3cb67431/mlir/include/mlir/IR/ValueRange.h#L374